ci: build-yocto: cleanup sstate-cache after the build

quaresmajose opened the pull request:
We need to clean up the ci local sstate-cache that is not used, otherwise it will grow infinitely.
This will take up significant space that will not be used by anyone since invalid sstate-cache
artefacts will no longer be needed.

Fixes https://github.com/qualcomm-linux/meta-qcom/issues/715

lumag commented on 0001-sstate-cache-management-dry-run.patch:
```diff
@@ -0,0 +1,34 @@
+From f28a08f27ca26579f2f5fa56ae52365fb16b8a32 Mon Sep 17 00:00:00 2001
+From: Jose Quaresma <jose.quaresma@foundries.io>
+Date: Mon, 31 Mar 2025 11:15:12 +0100
+Subject: [PATCH] sstate-cache-management: dry-run
```
It might be a good idea to add an actual `--dry-run` to the script.

quaresmajose replied:
Patch sent https://lore.kernel.org/openembedded-core/20250401165912.1603-1-jose.quaresma@foundries.io/T/#u

doanac commented on .github/workflows/build-yocto.yml:
```diff
@@ -117,6 +117,13 @@ jobs:
+          export SSTATE_DIR=${CACHE_DIR}/sstate-cache
+          export KAS_WORK_DIR=$PWD/../kas
+          SSTATE_CACHE_MANAGEMENT="${KAS_WORK_DIR}/poky/scripts/sstate-cache-management.py --cache-dir ${SSTATE_DIR} --remove-duplicated --remove-orphans --debug --yes"
+          kas shell ci/${{ matrix.machine }}.yml --command "$SSTATE_CACHE_MANAGEMENT"
```
not sure how long this script will take. My past recollection makes me think this will be pretty slow. Have you considered doing this on a cron schedule instead so that it doesn't affect day-to-day stuff as much?

ndechesne replied:
Yocto docs about sstate-cache-management says 

> It has the ability to purge all but the newest cache files on each architecture, and also to remove files that it considers unreachable by exploring a set of build configurations. However, this command requires a full build environment to be available and doesn’t work well covering multiple releases.

Which is not very encouraging to be honest..

quaresmajose replied:
for this first run it takes about 15 minutes for each machine but note that we have a very big sstate folder since it has never been cleaned yet. I expect to get much less time after this first interaction

ndechesne commented:
We should run it on push event only, not PR. First it would run less often, and then we really care about sstate for metadata which is merged, not what's in a PR which is temporary.

quaresmajose commented:
> We should run it on push event only, not PR. First it would run less often, and then we really care about sstate for metadata which is merged, not what's in a PR which is temporary.

God point! I have changed it to just run on git push events.

quaresmajose commented:
> When I looked at clean-sstate-cache a few years ago, it would remove everything unrelated to your existing build, so it would delete sstate for all other machines. Does it still do this or has it become smarter?

I believe the script is smart enough to support builds with more than one machine but I'm sure it will be like that. I will need to test this locally on my machine to actually see what it deletes.

lumag commented:
> We should run it on push event only, not PR. First it would run less often, and then we really care about sstate for metadata which is merged, not what's in a PR which is temporary.

I'd go for a scheduled cleanup, e.g. once a week or so.

ndechesne commented:
that would indeed work better. even once a month.. but that also means we need to duplicate the build config/machines in that cleanup job, since we need to inherit the same matrix config, right?

lumag commented:
Can't we define machines in a common config file which gets included by both workflows?

quaresmajose committed:
workflows/compile: cleanup sstate-cache

We need to clean up the CI sstate-cache that is not used, otherwise it
will grow infinitely. This will take up significant space that will not be
used by any build since invalid sstate-cache artifact will no longer be
needed.

Since all jobss run simultaneously, we need some kind of lock to ensure
that only one sstate-cache-management runs at a time. We use the flock
tool for that propose.

Signed-off-by: Jose Quaresma <jose.quaresma@oss.qualcomm.com>
